### PR TITLE
Improve message timestamp formatting.

### PR DESCRIPTION
**Pull request by charlesmchen-signal**
PTAL @michaelkirk

### CODE REVIEW

File: Signal/src/util/DateUtil.m [hunk 1]
```diff
@@ -357,46 +357,50 @@ + (NSString *)formatMessageTimestamp:(uint64_t)timestamp
 
     NSCalendar *calendar = [NSCalendar currentCalendar];
 
+    NSDateComponents *relativeDiffComponents =
```
Comment (charlesmchen-signal) [inline Signal/src/util/DateUtil.m]: Calculate the "relative time difference" minutes and hours as a single calculation, not as two separate calculations.

File: Signal/src/util/DateUtil.m [hunk 2]
```diff
@@ -357,46 +357,50 @@ + (NSString *)formatMessageTimestamp:(uint64_t)timestamp
...
+        [calendar components:NSCalendarUnitMinute | NSCalendarUnitHour fromDate:date toDate:nowDate options:0];
+
+    NSInteger minutesDiff = MAX(0, [relativeDiffComponents minute]);
+    if (minutesDiff < 1) {
+        return NSLocalizedString(@"DATE_NOW", @"The present; the current time.");
+    }
```
Comment (charlesmchen-signal) [inline Signal/src/util/DateUtil.m]: Exit early in the "now" case.

File: Signal/src/util/DateUtil.m [hunk 3]
```diff
@@ -357,46 +357,50 @@ + (NSString *)formatMessageTimestamp:(uint64_t)timestamp
...
+
+    NSInteger hoursDiff = MAX(0, [relativeDiffComponents hour]);
+    if (hoursDiff < 1) {
+        NSString *minutesString = [OWSFormat formatInt:(int)minutesDiff];
+        return [NSString stringWithFormat:NSLocalizedString(@"DATE_MINUTES_AGO_FORMAT",
+                                              @"Format string for a relative time, expressed as a certain number of "
+                                              @"minutes in the past. Embeds {{The number of minutes}}."),
+                         minutesString];
+    }
```
Comment (charlesmchen-signal) [inline Signal/src/util/DateUtil.m]: Exit early in the "N minutes ago" ago case.

File: Signal/src/util/DateUtil.m [hunk 4]
```diff
@@ -357,46 +357,50 @@ + (NSString *)formatMessageTimestamp:(uint64_t)timestamp
...
     if (yearsDiff > 0) {
         // "Long date" + locale-specific "short" time format.
         NSString *dayOfWeek = [self.otherYearMessageFormatter stringFromDate:date];
         NSString *formattedTime = [[self timeFormatter] stringFromDate:date];
-        result = [[dayOfWeek rtlSafeAppend:@" "] rtlSafeAppend:formattedTime];
-    } else if (daysDiff >= 7) {
+        return [[dayOfWeek rtlSafeAppend:@" "] rtlSafeAppend:formattedTime];
+    }
+
+    NSInteger daysDiff = [self daysFromFirstDate:date toSecondDate:nowDate];
```
Comment (charlesmchen-signal) [inline Signal/src/util/DateUtil.m]: Only calculate `daysDiff` if necessary.

File: Signal/src/util/DateUtil.m (hunk 2 above)
Comment (charlesmchen-signal) [reply Signal/src/util/DateUtil.m]: Note that this has perf benefits _and_ is a design refinement.   If someone sends or receives a message at Monday 11:59pm and looks at it at Tuesday 12:01 am, we want to see "2 minutes ago", not "Monday".  Myles has confirmed.

Comment (michaelkirk-signal) [reply Signal/src/util/DateUtil.m]: 👍

**Pull request merged**